Step2 문자열 계산기

serandev opened the pull request:
리뷰 부탁드립니다 😊

kakao-seren-a committed:
study: add kotlin basic

kakao-seren-a committed:
study: add kotlin basic kotest

kakao-seren-a committed:
feat: 자동 계산기 구현

- StringCalculator 로 요구사항 구현
- 정규식으로 입력 인자 유효성 검사
- 각 사칙연산 담당하는 Calculator 클래스 분리
- 문자열 식을 구현한 Expression 클래스 정의 및 식 계산 실행로직 구현

malibinYun commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,32 @@
+
+abstract class Calculator {
+    abstract fun calculate(firstNum: Int, secondNum: Int): Int
+}
```
다형성을 활용해서 계산기를 만드셨네요!
추상클래스 또는 인터페이스를 활용해서 서로 다른 구현체를 만드는 것은 좋은 접근이지만, when절 같은 분기 처리를 할 때는 휴먼 에러가 발생할 수 있습니다. 🙂
이럴떄는 Enum class 를 활용하거나, sealed class 또는 sealed interface를 활용하는 것이 좋습니다.
구현체가 새로 생기는 경우 when 구문에서 컴파일러가 감지하고 프로그래머에게 알려주기 때문이에요!

malibinYun replied:
덧붙여서, Calculator라는 네이밍은 StringCalculator와 혼동이 있을 수 있어 보여요.
StringCalculator도 Calculator의 구현체로 오해를 받을 수 있을거란 생각이 드네요.

"계산기" 보다는 계산을 수행하는 "연산자"라는 이름을 붙이는 것은 어떨까요?

serandev replied:
계산 로직이 단순해서 enum class 로 변경해봤습니다.
연산자인 operator 로 사용하면 코틀린 키워드와 겹칠 거 같아서 ArithmeticOperation 로 구체적인 의미가 담기도록 변경했어요

malibinYun commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,40 @@
+package step2
+
+class StringCalculator {
+    companion object {
```
내부 매서드들이 모두 companion object 안에 들어있고, 모두 순수 함수여서 object 키워드를 사용해도 좋아 보입니다.
```suggestion
object StringCalculator {
    ...
```

serandev replied:
이 방법을 모르고 있었네요 적용했습니다 👍

malibinYun commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,40 @@
+                val calculator = parseExpression(formulaStr[i - 1])
+                expression = Expression(number = number, calculator = calculator, nextExpression = expression)
+            }
+            return expression
```
문자열 계산기가 수식을 파싱하는 일도 하고있네요!
파싱하는 일은 다른 객체가 수행하게 만들어보는 것은 어떨까요?

serandev replied:
말씀주신대로 클래스를 분리하여 파싱하는 로직 이동해보았습니다
근데 Expression 클래스에 첫 번째 인자는 계산식 실행 시 넘겨줘야 되다보니 완전하게 분리해도 되는 클래스가 아닌 것 같아서 StringCalculator 클래스에 종속되는 것 같은데 고민이 되네요

malibinYun commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,40 @@
+            val expressionStr = input.replace(Regex("\\s+"), "")
+            if (expressionStr.isEmpty() || FORMULA_REGEX.matchEntire(expressionStr) == null) {
+                throw IllegalArgumentException()
+            }
```
Exception을 던질 때는 어떤 원인으로 발생한 것인지 메시지를 구체적으로 써주는 것이 좋습니다.
실제로 Exception이 발생했을 때 어떤 원인인지 콘솔창에서 찾고 디버깅하기 쉬워지기 때문이에요 🙂
```suggestion
                throw IllegalArgumentException("구체적인 메시지를 작성해주세요!")
            }
```

serandev replied:
메시지가 담기는게 좋을 거 같아서 말씀주신대로 수정했습니다

malibinYun commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,40 @@
+                '*' -> MultipleCalculator()
+                '/' -> DivisionCalculator()
+                else -> throw IllegalArgumentException()
+            }
```
연산자를 찾는 일은, 연산자가 갖고있게 하는 것은 어떨까요?

serandev replied:
해당 클래스로 로직 이동했습니다

malibinYun commented on src/test/kotlin/step2/CalculatorTest.kt:
```diff
@@ -0,0 +1,70 @@
+
+        val expected = firstNumber + secondNumber
+        assertThat(result).isEqualTo(expected)
+    }
```
테스트 코드의 검증부는 하드 코딩하는 것이 좋습니다.
PlusCalculator 내부의 로직이 + 연산이 아닌 다른 종의 것으로 리팩터링 된다면, 테스트코드의 expected 로직은 그냥 두어야할지, 도메인 로직을 다시 복사할지 난감한 상황이 발생하게 됩니다.

검증부 하드 코딩에 대한 좋은 글을 공유해드릴게요. 꼭 읽어보세요~
https://jojoldu.tistory.com/615?category=1036934

serandev replied:
좋은 글 공유 감사합니다 👍 
비슷하게 사용했던 테스트코드들 수정했어요

malibinYun commented on src/test/kotlin/step2/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,35 @@
+
+    @ParameterizedTest
+    @CsvSource("2 + 3 ? 4", "2 +", "2 + 3 * 4 /", "+")
+    fun `식 형태 오류`(input: String) {
```
오류인 형태의 식 검증은 더 세밀하게 나누어도 좋다는 생각 이듭니다.
알 수 없는 연산자가 들어왔을 때, 미완성인 수식 등 각 Exception이 다르고 더 세밀한 핸들링을 할 수 있겠네요 🙂

malibinYun commented on src/test/kotlin/study/PersonKoTest.kt:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+
+    "널 타입" {
+        val person = Person("염세란", 32, nickname = null)
+        person.name shouldBe "염세란"
+        person.age shouldBe 32
+        person.nickname shouldBe "세렌"
+    }
```
요 테스트는 실패하고 있네요!

serandev replied:
놓쳤었네요;; 수정했습니다

malibinYun reviewed (changes requested):
전반적으로 설계가 좋았다는 생각이 듭니다 :)
생각해보면 좋을 부분에 코멘트 남겨두었으니 리뷰 반영해주세요!

kakao-seren-a committed:
refactor: 코드 리뷰 적용
- StringCalculator 내에서 문자열 파싱 로직 분리
- StringCalculator 네이밍 변경
- Calculator 네이밍 변경 및 abstract -> enum class 변경
- IllegalArgumentException 에러메시지 명시
- 이전 단계 테스트코드 오류 수정

serandev committed:
Merge branch 'sereran' into step2

malibinYun commented on src/main/kotlin/step2/ArithmeticOperation.kt:
```diff
@@ -0,0 +1,38 @@
+package step2
+
+enum class ArithmeticOperation(
+    private val expression: Char
+) {
+    PLUS('+') {
+        override fun execute(firstNum: Int, secondNum: Int): Int {
+            return firstNum + secondNum
+        }
+    },
```
enum 클래스의 멤버 변수로 람다를 갖게 한다면 더 깔끔하게 작성하실 수 있습니다 🙂

```suggestion
enum class ArithmeticOperation(
    private val expression: Char,
    val calculate: (Int, Int) -> Int
) {
    PLUS('+', { firstNum, secondNum -> firstNum + secondNum }),
```

malibinYun commented on src/main/kotlin/step2/StringExpressionParser.kt:
```diff
@@ -0,0 +1,29 @@
+
+    private fun parseFormula(formulaStr: String): Expression? {
+        var expression: Expression? = null
+        for (i: Int in formulaStr.length - 1 downTo 1) {
+            if (i % 2 == 1) {
+                continue
+            }
```
두 단계씩 i 값을 변경시키는 것이 의도이시라면, 아래 코틀린 기능이 도움이 되실겁니다 :)
```suggestion
        for(i in formulaStr.length - 1 downTo 1 step 2)
```

malibinYun commented on src/main/kotlin/step2/ArithmeticOperation.kt:
```diff
@@ -0,0 +1,38 @@
+    companion object {
+        fun findOperation(expression: Char): ArithmeticOperation {
+            return values().firstOrNull { expression == it.expression }
+                ?: throw IllegalArgumentException("$expression 에 해당되는 연산자를 찾을 수 없습니다.")
+        }
```
Exception이 발생하면 메시지로 버그 추적이 쉬워지겠네요! 👍

malibinYun reviewed (approved):
피드백 반영 고생 많으셨어요!
다음 미션을 위해 머지하겠습니다.
생각해보면 좋을 점들에 코멘트 달아두었으니 확인해주세요!
